lib/Model/Pass/FlattenPrimitiveTypedefs.cpp

```diff
@@ -0,0 +1,57 @@
+/// \file FlattenPrimitiveTypedefs.cpp
+/// Implementation of deduplication of colliding names.
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Binary.h"
+#include "revng/Model/Pass/FlattenPrimitiveTypedefs.h"
+#include "revng/Model/Pass/RegisterModelPass.h"
+#include "revng/Model/Processing.h"
+
+static RegisterModelPass R("flatten-primitive-typedefs",
+                           "This looks for typedefs of common primitive types "
+                           "and replaces them with their type.",
+                           model::flattenPrimitiveTypedefs);
+
+void model::flattenPrimitiveTypedefs(TupleTree<model::Binary> &Binary) {
```

Comment:
1. This pass also runs on non-valid models, correct?
2. I think this should then become part of the "model::fixModel" thing.

Reply:
@ale, I'll continue the other thread here:

https://github.com/revng/revng/pull/432#discussion_r1969926692

Since there seems to be fundamental misunderstanding of why we need a more powerful tool for iterating all the types.

> Not sure what you mean. 🤔

The point of this pass is to straight-up remove typedefs with names that conflict with known primitive types (as long as sized match). Hence it needs to be able to _remove_ types, which is not possible over a simple `edges()` iteration.

Reply:
> It seems to me that we're assuming the typedef name is something like `uint64_t`:

that's the point. Dwarf gives us those, and because of the name collision, they end up being shown as `typedef_XXX` which is ugly.

Reply:
> No this pass runs on valid models.

And...

> > It seems to me that we're assuming the typedef name is something like uint64_t:
> Dwarf gives us those

A model with a typedef called `uint64_t` is not a valid model.
But OK.

Are you aware of `TupleTree:replaceReferences` and friends?

Reply:
> A model with a typedef called `uint64_t` is not a valid model.

Wait, it is now! Since the typedef is just going to get renamed to its "default" name. So it's never checked as part of verification anymore.

> Are you aware of `TupleTree:replaceReferences` and friends?

They only allow to modify references, and here we _are_ removing them. Specifically by getting rid of `DefinedType`s (and references they contain) and inserting `PrimitiveType`s instead.

Reply:
Ah, right, I didn't think about going that deep into tuple-tree guts. But yeah, that should be doable with an inheritor of `VisitorBase`.

Wait, on that note, I wonder if we should reimplement `edges` through it as well while at it 🤔. The problem of having to list fields manually is not exclusive to this new thing.

That being said, good catch!

Reply:
It might be possible, yeah.

My main concern right now is to minimize the places where we list all the model parts (e.g., inheritors of `TypeDefinitions` and their fields) around the code base. It's a liability when we make changes to the model.

`edges` for now it's within the model itself, so it's less risky.

Reply:
Actually this turned out to be surprisingly easy, sorry for not looking into it deeply enough intially.